lbrynet/lbrynet_daemon/auth/server.py

```diff
@@ -0,0 +1,193 @@
+import logging
+
+from decimal import Decimal
+from twisted.web import server
+from twisted.internet import defer
+from txjsonrpc import jsonrpclib
+from txjsonrpc.web import jsonrpc
+from txjsonrpc.web.jsonrpc import Handler
+
+from lbrynet.core.Error import InvalidAuthenticationToken, InvalidHeaderError
+from lbrynet.lbrynet_daemon.auth.util import APIKey
+from lbrynet.conf import ALLOWED_DURING_STARTUP
+
+log = logging.getLogger(__name__)
+
+
+def default_decimal(obj):
+    if isinstance(obj, Decimal):
+        return float(obj)
+
+
+def authorizer(cls):
+    cls.authorized_functions = []
+    for methodname in dir(cls):
+        if methodname.startswith("jsonrpc_"):
+            method = getattr(cls, methodname)
+            if hasattr(method, '_auth_required'):
+                cls.authorized_functions.append(methodname.split("jsonrpc_")[1])
+    return cls
+
+
+def auth_required(f):
+    f._auth_required = True
+    return f
+
+
+@authorizer
+class LBRYJSONRPCServer(jsonrpc.JSONRPC):
+
+    isLeaf = True
+
+    def __init__(self):
+        jsonrpc.JSONRPC.__init__(self)
+        self.sessions = {}
+
+    def _register_user_session(self, session_id):
+        token = APIKey.new()
+        self.sessions.update({session_id: token})
+        return token
+
+    def _responseFailed(self, err, call):
+        log.debug(err.getTraceback())
+
+    def _set_headers(self, request, data):
+        request.setHeader("Access-Control-Allow-Origin", "localhost")
+        request.setHeader("Content-Type", "text/json")
+        request.setHeader("Content-Length", str(len(data)))
+
+    def _render_message(self, request, message):
+        request.write(message)
+        request.finish()
+
+    def _check_headers(self, request):
+        origin = request.getHeader("Origin")
+        referer = request.getHeader("Referer")
+
+        if origin not in [None, 'http://localhost:5279']:
+            log.warning("Attempted api call from %s", origin)
+            raise InvalidHeaderError
+
+        if referer is not None and not referer.startswith('http://localhost:5279/'):
```

Comment:
This may prevent someone from opening an lbry:// link from another site. You should test this